File: ruby/test/data_structures_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../lib/data_structures/stack'
require_relative '../lib/data_structures/queue'
require_relative '../lib/data_structures/dynamic_array'
require_relative '../lib/data_structures/singly_linked_list'
require_relative '../lib/data_structures/doubly_linked_list'
require_relative '../lib/data_structures/deque'
require_relative '../lib/data_structures/hash_table'
require_relative '../lib/data_structures/binary_search_tree'
require_relative '../lib/data_structures/min_heap'
require_relative '../lib/data_structures/disjoint_set'

class DataStructuresTest < Minitest::Test
  # Stack Tests
  def test_stack_operations
    stack = DSA::DataStructures::Stack.new
    assert stack.empty?

    stack.push(1).push(2).push(3)

    assert_equal 3, stack.size
    assert_equal 3, stack.peek
    assert_equal 3, stack.pop
    assert_equal 2, stack.pop
    assert_equal 1, stack.size

    stack.clear
    assert stack.empty?
  end

  # Queue Tests
  def test_queue_operations
    queue = DSA::DataStructures::Queue.new
    assert queue.empty?

    queue.enqueue(1).enqueue(2).enqueue(3)

    assert_equal 3, queue.size
    assert_equal 1, queue.peek
    assert_equal 1, queue.dequeue
    assert_equal 2, queue.dequeue
    assert_equal 1, queue.size
  end

  # DynamicArray Tests
  def test_dynamic_array_operations
    arr = DSA::DataStructures::DynamicArray.new
    assert arr.empty?

    arr.push(10).push(20).push(30)

    assert_equal 3, arr.size
    assert_equal 10, arr.get(0)
    assert_equal 30, arr.get(2)

    assert arr.set(1, 25)
    assert_equal 25, arr.get(1)

    assert_equal 30, arr.pop
    assert_equal 2, arr.size

    assert arr.insert(1, 15)
    assert_equal [10, 15, 25], arr.to_a

    assert_equal 15, arr.remove_at(1)
    assert_equal [10, 25], arr.to_a
  end

  # SinglyLinkedList Tests
  def test_singly_linked_list_operations
    list = DSA::DataStructures::SinglyLinkedList.new
    assert list.empty?

    list.append(1).append(2).prepend(0)

    assert_equal 3, list.size
    assert_equal 0, list.get(0)
    assert_equal 2, list.get(2)

    assert list.insert(1, 10)
    assert_equal [0, 10, 1, 2], list.to_a

    assert_equal 10, list.remove_at(1)
    assert_equal [0, 1, 2], list.to_a

    assert list.contains?(1)
    refute list.contains?(100)
  end

  # DoublyLinkedList Tests
  def test_doubly_linked_list_operations
    list = DSA::DataStructures::DoublyLinkedList.new
    assert list.empty?

    list.append(1).append(2).prepend(0)

    assert_equal 3, list.size
    assert_equal 0, list.first
    assert_equal 2, list.last

    assert_equal 0, list.remove_first
    assert_equal 2, list.remove_last
    assert_equal 1, list.size
  end

  # Deque Tests
  def test_deque_operations
    deque = DSA::DataStructures::Deque.new
    assert deque.empty?

    deque.push_back(2).push_front(1).push_back(3)

    assert_equal 3, deque.size
    assert_equal 1, deque.peek_front
    assert_equal 3, deque.peek_back

    assert_equal 1, deque.pop_front
    assert_equal 3, deque.pop_back
    assert_equal 1, deque.size
  end

  # HashTable Tests
  def test_hash_table_operations
    table = DSA::DataStructures::HashTable.new
    assert table.empty?

    table.put('one', 1).put('two', 2).put('three', 3)

    assert_equal 3, table.size
    assert_equal 1, table.get('one')
    assert_equal 2, table.get('two')
    assert_nil table.get('four')

    assert table.contains?('one')
    refute table.contains?('four')

    table.put('one', 100)
    assert_equal 100, table.get('one')

    assert_equal 2, table.remove('two')
    assert_equal 2, table.size
    refute table.contains?('two')
  end

  # BinarySearchTree Tests
  def test_binary_search_tree_operations
    bst = DSA::DataStructures::BinarySearchTree.new
    assert bst.empty?

    bst.insert(5).insert(3).insert(7).insert(1).insert(9)

    assert_equal 5, bst.size
    assert bst.contains?(5)
    assert bst.contains?(3)
    refute bst.contains?(100)

    assert_equal 1, bst.min
    assert_equal 9, bst.max

    assert_equal [1, 3, 5, 7, 9], bst.in_order

    assert bst.remove(3)
    refute bst.contains?(3)
    assert_equal [1, 5, 7, 9], bst.in_order
  end

  # MinHeap Tests
  def test_min_heap_operations
    heap = DSA::DataStructures::MinHeap.new
    assert heap.empty?

    heap.insert(5).insert(3).insert(7).insert(1).insert(9)

    assert_equal 5, heap.size
    assert_equal 1, heap.peek

    assert_equal 1, heap.extract_min
    assert_equal 3, heap.extract_min
    assert_equal 5, heap.extract_min
    assert_equal 7, heap.extract_min
    assert_equal 9, heap.extract_min

    assert heap.empty?
  end

  def test_min_heap_with_custom_comparator
    # Max heap using custom comparator
    heap = DSA::DataStructures::MinHeap.new { |a, b| b <=> a }

    heap.insert(5).insert(3).insert(7)

    assert_equal 7, heap.extract_min
    assert_equal 5, heap.extract_min
    assert_equal 3, heap.extract_min
  end

  # DisjointSet Tests
  def test_disjoint_set_operations
    ds = DSA::DataStructures::DisjointSet.new(5)

    assert_equal 5, ds.count
    refute ds.connected?(0, 1)

    assert ds.union(0, 1)
    assert ds.connected?(0, 1)
    assert_equal 4, ds.count

    assert ds.union(2, 3)
    assert ds.union(0, 2)

    assert ds.connected?(0, 3)
    assert ds.connected?(1, 2)
    assert_equal 2, ds.count

    refute ds.connected?(0, 4)
  end
end

File: scala/src/test/scala/DataStructuresTest.scala
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers
import datastructures.*

class DataStructuresTest extends AnyFunSuite with Matchers:

  // Stack Tests
  test("Stack - push and pop"):
    val stack = Stack[Int]()
    stack.push(1).push(2).push(3)
    stack.pop() shouldBe Some(3)
    stack.pop() shouldBe Some(2)
    stack.pop() shouldBe Some(1)
    stack.pop() shouldBe None

  test("Stack - peek"):
    val stack = Stack[Int]()
    stack.push(42)
    stack.peek shouldBe Some(42)
    stack.size shouldBe 1

  test("Stack - isEmpty"):
    val stack = Stack[Int]()
    stack.isEmpty shouldBe true
    stack.push(1)
    stack.isEmpty shouldBe false

  // Queue Tests
  test("Queue - enqueue and dequeue"):
    val queue = Queue[Int]()
    queue.enqueue(1).enqueue(2).enqueue(3)
    queue.dequeue() shouldBe Some(1)
    queue.dequeue() shouldBe Some(2)
    queue.dequeue() shouldBe Some(3)
    queue.dequeue() shouldBe None

  test("Queue - peek"):
    val queue = Queue[Int]()
    queue.enqueue(42)
    queue.peek shouldBe Some(42)
    queue.size shouldBe 1

  // DynamicArray Tests
  test("DynamicArray - push and pop"):
    val arr = DynamicArray[Int]()
    arr.push(1).push(2).push(3)
    arr.pop() shouldBe Some(3)
    arr.size shouldBe 2

  test("DynamicArray - get and set"):
    val arr = DynamicArray[Int]()
    arr.push(1).push(2).push(3)
    arr.get(1) shouldBe Some(2)
    arr.set(1, 42) shouldBe true
    arr.get(1) shouldBe Some(42)

  test("DynamicArray - insert and removeAt"):
    val arr = DynamicArray[Int]()
    arr.push(1).push(3)
    arr.insert(1, 2) shouldBe true
    arr.toList shouldBe List(1, 2, 3)
    arr.removeAt(1) shouldBe Some(2)
    arr.toList shouldBe List(1, 3)

  // SinglyLinkedList Tests
  test("SinglyLinkedList - prepend and append"):
    val list = SinglyLinkedList[Int]()
    list.prepend(2).prepend(1).append(3)
    list.toList shouldBe List(1, 2, 3)

  test("SinglyLinkedList - get and indexOf"):
    val list = SinglyLinkedList[Int]()
    list.append(10).append(20).append(30)
    list.get(1) shouldBe Some(20)
    list.indexOf(20) shouldBe 1
    list.indexOf(100) shouldBe -1

  test("SinglyLinkedList - insert and removeAt"):
    val list = SinglyLinkedList[Int]()
    list.append(1).append(3)
    list.insert(1, 2) shouldBe true
    list.toList shouldBe List(1, 2, 3)
    list.removeAt(1) shouldBe Some(2)
    list.toList shouldBe List(1, 3)

  // DoublyLinkedList Tests
  test("DoublyLinkedList - prepend and append"):
    val list = DoublyLinkedList[Int]()
    list.prepend(2).prepend(1).append(3)
    list.toList shouldBe List(1, 2, 3)

  test("DoublyLinkedList - first and last"):
    val list = DoublyLinkedList[Int]()
    list.append(1).append(2).append(3)
    list.first shouldBe Some(1)
    list.last shouldBe Some(3)

  test("DoublyLinkedList - removeFirst and removeLast"):
    val list = DoublyLinkedList[Int]()
    list.append(1).append(2).append(3)
    list.removeFirst() shouldBe Some(1)
    list.removeLast() shouldBe Some(3)
    list.toList shouldBe List(2)

  // Deque Tests
  test("Deque - pushFront and pushBack"):
    val deque = Deque[Int]()
    deque.pushFront(2).pushFront(1).pushBack(3)
    deque.popFront() shouldBe Some(1)
    deque.popBack() shouldBe Some(3)
    deque.popFront() shouldBe Some(2)

  test("Deque - peekFront and peekBack"):
    val deque = Deque[Int]()
    deque.pushBack(1).pushBack(2)
    deque.peekFront shouldBe Some(1)
    deque.peekBack shouldBe Some(2)

  // HashTable Tests
  test("HashTable - put and get"):
    val table = HashTable[String, Int]()
    table.put("one", 1).put("two", 2).put("three", 3)
    table.get("one") shouldBe Some(1)
    table.get("two") shouldBe Some(2)
    table.get("four") shouldBe None

  test("HashTable - remove"):
    val table = HashTable[String, Int]()
    table.put("key", 42)
    table.remove("key") shouldBe Some(42)
    table.get("key") shouldBe None

  test("HashTable - keys and values"):
    val table = HashTable[String, Int]()
    table.put("a", 1).put("b", 2)
    table.keys.toSet shouldBe Set("a", "b")
    table.values.toSet shouldBe Set(1, 2)

  test("HashTable - resize"):
    val table = HashTable[Int, Int](4)
    for i <- 1 to 20 do table.put(i, i * 10)
    table.size shouldBe 20
    for i <- 1 to 20 do table.get(i) shouldBe Some(i * 10)

  // BinarySearchTree Tests
  test("BST - insert and contains"):
    val bst = BinarySearchTree[Int]()
    bst.insert(5).insert(3).insert(7).insert(1).insert(9)
    bst.contains(5) shouldBe true
    bst.contains(3) shouldBe true
    bst.contains(10) shouldBe false

  test("BST - min and max"):
    val bst = BinarySearchTree[Int]()
    bst.insert(5).insert(3).insert(7).insert(1).insert(9)
    bst.min shouldBe Some(1)
    bst.max shouldBe Some(9)

  test("BST - remove"):
    val bst = BinarySearchTree[Int]()
    bst.insert(5).insert(3).insert(7)
    bst.remove(5) shouldBe true
    bst.contains(5) shouldBe false
    bst.size shouldBe 2

  test("BST - inorder traversal"):
    val bst = BinarySearchTree[Int]()
    bst.insert(5).insert(3).insert(7).insert(1).insert(9)
    bst.inorder shouldBe List(1, 3, 5, 7, 9)

  // MinHeap Tests
  test("MinHeap - insert and extractMin"):
    val heap = MinHeap[Int]()
    heap.insert(5).insert(3).insert(7).insert(1).insert(9)
    heap.extractMin() shouldBe Some(1)
    heap.extractMin() shouldBe Some(3)
    heap.extractMin() shouldBe Some(5)

  test("MinHeap - peek"):
    val heap = MinHeap[Int]()
    heap.insert(5).insert(3).insert(7)
    heap.peek shouldBe Some(3)
    heap.size shouldBe 3

  test("MinHeap - from factory"):
    val heap = MinHeap.from(List(5, 3, 7, 1, 9))
    heap.extractMin() shouldBe Some(1)
    heap.extractMin() shouldBe Some(3)

  // DisjointSet Tests
  test("DisjointSet - makeSet and find"):
    val ds = DisjointSet[Int]()
    ds.makeSet(1).makeSet(2).makeSet(3)
    ds.find(1) shouldBe Some(1)
    ds.find(4) shouldBe None

  test("DisjointSet - union and connected"):
    val ds = DisjointSet[Int]()
    ds.makeSet(1).makeSet(2).makeSet(3)
    ds.connected(1, 2) shouldBe false
    ds.union(1, 2) shouldBe true
    ds.connected(1, 2) shouldBe true
    ds.connected(1, 3) shouldBe false

  test("DisjointSet - setCount"):
    val ds = DisjointSet[Int]()
    ds.makeSet(1).makeSet(2).makeSet(3).makeSet(4)
    ds.setCount shouldBe 4
    ds.union(1, 2)
    ds.setCount shouldBe 3
    ds.union(3, 4)
    ds.setCount shouldBe 2
    ds.union(1, 3)
    ds.setCount shouldBe 1
